code/Controllers/LeftAndMainPageIconsExtension.php

```diff
@@ -32,28 +35,25 @@ public function generatePageIconsCss()
 
         $classes = ClassInfo::subclassesFor(SiteTree::class);
         foreach ($classes as $class) {
-            $obj = singleton($class);
-            $iconSpec = $obj->config()->get('icon');
-
-            if (!$iconSpec) {
+            $icon = Config::inst()->get($class, 'icon');
+            if (!$icon) {
                 continue;
             }
 
-            // Legacy support: We no longer need separate icon definitions for folders etc.
-            $iconFile = (is_array($iconSpec)) ? $iconSpec[0] : $iconSpec;
-
-            // Legacy support: Add file extension if none exists
-            if (!pathinfo($iconFile, PATHINFO_EXTENSION)) {
-                $iconFile .= '-file.gif';
-            }
-
             $class = Convert::raw2htmlid($class);
             $selector = ".page-icon.class-$class, li.class-$class > a .jstree-pageicon";
-            if (Director::fileExists($iconFile)) {
-                $css .= "$selector { background: transparent url('$iconFile') 0 0 no-repeat; }\n";
+            $iconResource = ModuleResourceLoader::singleton()->resolveResource($icon);
+            if ($iconResource instanceof ModuleResource) {
+                // Support module resource format
+                $iconURL = $iconResource->getURL();
+                $css .= "$selector { background: transparent url('$iconURL') 0 0 no-repeat; }\n";
+            } elseif (Director::fileExists($icon)) {
+                // Support direct path
+                $iconURL = ModuleResourceLoader::resourceURL($icon);
+                $css .= "$selector { background: transparent url('$iconURL') 0 0 no-repeat; }\n";
             } else {
                 // Support for more sophisticated rules, e.g. sprited icons
-                $css .= "$selector { $iconFile }\n";
+                $css .= "$selector { $icon }\n";
```

Comment:
should this be sanitised?
I'm a bit wary of someone injecting css code here

Reply:
I think we may need this in Requirements_Backend

```diff
// Literal custom CSS content
foreach ($this->getCustomCSS() as $css) {
+    $css = Convert::raw2xml($css);
    $requirements .= HTML::createTag('style', ['type' => 'text/css'], "\n{$css}\n");
    $requirements .= "\n";
}
```

But separate story to this.

Reply:
Looks like it's not that easy.

Possible answer https://stackoverflow.com/questions/3241616/sanitize-user-defined-css-in-php, but requires a thirdparty lib.

Reply:
Let's dump this risky behaviour. It's a risk identified on owasp so let's at least minimise it https://www.owasp.org/index.php/XSS_(Cross_Site_Scripting)_Prevention_Cheat_Sheet#HTML_entity_encoding